cpp/test/neighbors/ann_mg.cuh

```diff
@@ -0,0 +1,450 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "../test_utils.cuh"
+#include "ann_utils.cuh"
+#include "naive_knn.cuh"
+
+#include <cuvs/neighbors/ann_mg.hpp>
+
+namespace cuvs::neighbors::mg {
+
+struct AnnMGInputs {
+  int64_t num_queries;
+  int64_t num_db_vecs;
+  int64_t dim;
+  int64_t k;
+  int64_t nprobe;
+  int64_t nlist;
+  cuvs::distance::DistanceType metric;
+  bool adaptive_centers;
+};
+
+template <typename T, typename DataT>
+class AnnMGTest : public ::testing::TestWithParam<AnnMGInputs> {
+ public:
+  AnnMGTest()
+    : stream_(resource::get_cuda_stream(handle_)),
+      ps(::testing::TestWithParam<AnnMGInputs>::GetParam()),
+      d_index_dataset(0, stream_),
+      d_query_dataset(0, stream_),
+      h_index_dataset(0),
+      h_query_dataset(0)
+  {
+  }
+
+  void testAnnMG()
+  {
+    size_t queries_size = ps.num_queries * ps.k;
+    std::vector<int64_t> indices_naive(queries_size);
+    std::vector<T> distances_naive(queries_size);
+    std::vector<int64_t> indices_ann(queries_size);
+    std::vector<T> distances_ann(queries_size);
+    std::vector<uint32_t> indices_naive_32bits(queries_size);
+    std::vector<uint32_t> indices_ann_32bits(queries_size);
+
+    {
+      rmm::device_uvector<T> distances_naive_dev(queries_size, stream_);
+      rmm::device_uvector<int64_t> indices_naive_dev(queries_size, stream_);
+      cuvs::neighbors::naive_knn<T, DataT, int64_t>(handle_,
+                                                    distances_naive_dev.data(),
+                                                    indices_naive_dev.data(),
+                                                    d_query_dataset.data(),
+                                                    d_index_dataset.data(),
+                                                    ps.num_queries,
+                                                    ps.num_db_vecs,
+                                                    ps.dim,
+                                                    ps.k,
+                                                    ps.metric);
+      update_host(distances_naive.data(), distances_naive_dev.data(), queries_size, stream_);
+      update_host(indices_naive.data(), indices_naive_dev.data(), queries_size, stream_);
+      resource::sync_stream(handle_);
+    }
+
+    int n_devices;
+    cudaGetDeviceCount(&n_devices);
+    std::cout << n_devices << " GPUs detected" << std::endl;
+    std::vector<int> device_ids(n_devices);
+    std::iota(device_ids.begin(), device_ids.end(), 0);
+
+    uint64_t n_rows_per_batch = 3000;  // [3000, 3000, 1000] == 7000 rows
+    cuvs::neighbors::mg::nccl_clique clique(device_ids);
+
+    // IVF-Flat
+    for (parallel_mode d_mode : {parallel_mode::REPLICATED, parallel_mode::SHARDED}) {
```

Comment:
This feels very verbose to manually repeat the test blocks for each algo. Additionally you are combining multiple test into a single test instance: parallel_modes x [vector search algo]. Ideally these should go into test params, define test params as tensor product of various options, and each test should run a single algo. This need not block the PR, but would be a nice to have.